tests/sub_group/sub_group_api.cpp

```diff
@@ -68,34 +68,106 @@ static size_t get_work_group_size(const sycl::device& device) {
       "Implementation does not define device.get_info<max_work_item_sizes>. "
       "Using work-group of size (1, 1, 1).");
 #endif
-  const size_t smallest_max_work_item_size =
-      std::min(max_work_item_sizes[0],
-               std::min(max_work_item_sizes[1], max_work_item_sizes[2]));
-  const size_t desired_work_group_size =
-      std::min<size_t>(smallest_max_work_item_size, 1024);
   const size_t max_work_group_size =
       device.get_info<sycl::info::device::max_work_group_size>();
 
-  // calc ilog2(desired_work_group_size), desired_work_group_size is at least 1
-  size_t tmp = desired_work_group_size;
-  size_t desired_work_group_size_exp = 0;
-  while (tmp >>= 1) desired_work_group_size_exp++;
-
-  size_t size = 1;
-  for (size_t curr_exp = 1; curr_exp < desired_work_group_size_exp;
-       curr_exp++) {
-    const size_t curr_size = static_cast<size_t>(1) << curr_exp;
-    if (curr_size <= desired_work_group_size &&
-        curr_size * curr_size * curr_size <= max_work_group_size) {
-      size = curr_size;
-    } else {
-      break;
+  std::vector<size_t> desired_work_group_sizes = {
+      std::min<size_t>(max_work_item_sizes[0], 1024),
+      std::min<size_t>(max_work_item_sizes[1], 1024),
+      std::min<size_t>(max_work_item_sizes[2], 1024)};
+
+  // calc ilog2(desired_work_group_sizes), desired_work_group_sizes is at least
+  // 1
+  std::vector<size_t> desired_work_group_sizes_exp(3);
+  auto calc_ilog2 = [](size_t dim_size) {
+    size_t tmp = dim_size;
+    size_t dim_size_exp = 0;
+    while (tmp >>= 1) dim_size_exp++;
+    return dim_size_exp;
+  };
+
+  std::transform(desired_work_group_sizes.begin(),
+                 desired_work_group_sizes.end(),
+                 desired_work_group_sizes_exp.begin(), calc_ilog2);
+
+  auto tot_work_items = [](std::vector<size_t> sizes) {
```

Comment:
```suggestion
  auto tot_work_items = [](const std::vector<size_t>& sizes) {
```
or even
```suggestion
  auto tot_work_items = [](const auto& sizes) {
```
since you are not checking the number of elements anyway.
But looking at the code, you could move this lambda in the following lambda or just inline it there since it is only used once... :-)